src/masternode-budget.cpp

```diff
@@ -1906,9 +1904,9 @@ CAmount CFinalizedBudget::GetTotalPayout() const
 
 std::string CFinalizedBudget::GetProposals()
 {
-    LOCK(cs);
-    std::string ret = "";
+    LOCK(budget.cs_proposals);
```

Comment:
Instead of locking a mutex from outside, would be better to create a method in `CBudgetManager` called `GetProposals(std::vector<uint256>)` and pass `vecBudgetPayments` proposal hashes, lock `cs_proposal` inside and return the string.

(Same applies for `CFinalizedBudget::GetStatus`. We should try to avoid locking a mutex from outside of the owner object)

Reply:
Yes, agree.. I didn't like this implementations either (but didn't want to change the flow too much, yet).
I think we can do what you suggest, and even go one step further: save the string of proposal names directly inside the CFinalizedBudget class.
This way we can do the task of retrieving the names of proposals (from their hashes) with the manager, only once, before adding the budget to the map (thus without the need to lock cs_budgets), and then access it easily.
I've tried this out in ae17fb612b5bc38a5e2348d3fd4bb649a496055a.

`GetStatus` is even uglier. I've moved it directly inside the manager in 59f30244448a55523fec7440e2ff882e3acce18d (after introducing constant getters for budget objects in the maps).

Reply:
looking much better 👌.

tiny thing in 89de05c. There is a really edge case scenario in `AddFinalizedBudget` when we don't lock `cs_budgets` over the entire method. decoupling `mapFinalizedBudgets.count` and `mapFinalizedBudgets.emplace` opens a small window where two equal proposals entering in the method will be executed.
We will not have any problems if this really unlikely scenario happens over this code, writing it to have it in mind for other areas as well.

Reply:
Yes, absolutely. I thought about that possibility, and considered two alternatives:
1- lock cs_budgets for the entire function (`AddFinalizedBudget`) scope
2- move the `UpdateValid` checks at the top, and lock cs_budgets only after (to check for existence, and add to the map)

(1) Presented the problem of invalid lock order, as the collateral check (`UpdateValid`) needs `cs_main`.
(2) Is inefficient, as we would perform the checks (holding the locks needed) even for budgets that we already have.

The solution of splitting `map.count` and `map.emplace` locking is harmless: if the same budgets are concurrently passed to this function, the second one that calls `map.emplace` is actually a no-op (it returns a `std::pair<iterator,bool>`, where `iterator` points the already-existing element in the map, and the bool is `false`, indicating that no insertion happened).